code/hsec-sync/src/Security/Advisories/Sync.hs

```diff
@@ -1,76 +1,95 @@
   | Updated
   | AlreadyUpToDate
   deriving stock (Eq, Show)
 
-sync :: Repository -> IO (Either String SyncStatus)
-sync repo = do
-  gitStatus <- gitRepositoryStatus repo
-  ensured <- ensureGitRepositoryWithRemote repo gitStatus
-  let mkGitError = Left . explainGitError
-  case ensured of
-    Left e -> return $ mkGitError e
-    Right s ->
-      case s of
-        GitRepositoryCreated ->
-          return $ Right Created
-        GitRepositoryExisting -> do
-          repoStatus <- status' repo gitStatus
-          if repoStatus == DirectoryOutDated
-            then either mkGitError (const $ Right Updated) <$> updateGitRepository repo
-            else return $ Right AlreadyUpToDate
+sync :: Snapshot -> IO (Either String SyncStatus)
+sync s =
+  runExceptT $ do
+    snapshotStatus <- liftIO $ snapshotRepositoryStatus $ snapshotRoot s
+    ensuredStatus <- withExceptT explainSnapshotError $ ensureSnapshot (snapshotRoot s) (snapshotUrl s) snapshotStatus
+    case ensuredStatus of
+      SnapshotRepositoryCreated ->
+        return Created
+      SnapshotRepositoryExisting -> do
+        repoStatus <- liftIO $ status' s snapshotStatus
+        if repoStatus == DirectoryOutDated
+          then do
+            withExceptT explainSnapshotError $ overwriteSnapshot (snapshotRoot s) (snapshotUrl s)
+            return Updated
+          else return AlreadyUpToDate
 
 data RepositoryStatus
   = DirectoryMissing
-  | DirectoryEmpty
+  | -- | Used when expected files/directories are missing or not readable
+    DirectoryIncoherent
   | DirectoryUpToDate
   | DirectoryOutDated
   deriving stock (Eq, Show)
 
-status :: Repository -> IO RepositoryStatus
-status repo =
-  status' repo =<< gitRepositoryStatus repo
+status :: Snapshot -> IO RepositoryStatus
+status s =
+  status' s =<< snapshotRepositoryStatus (snapshotRoot s)
 
-status' :: Repository -> GitRepositoryStatus -> IO RepositoryStatus
-status' repo gitStatus = do
-  case gitStatus of
-    GitDirectoryMissing ->
+status' :: Snapshot -> SnapshotRepositoryStatus -> IO RepositoryStatus
+status' s =
+  \case
+    SnapshotDirectoryMissing ->
       return DirectoryMissing
-    GitDirectoryEmpty ->
-      return DirectoryEmpty
-    GitDirectoryInitialized -> do
-      gitInfo <- getDirectoryGitInfo $ repositoryRoot repo
-      case gitInfo of
+    SnapshotDirectoryIncoherent ->
+      return DirectoryIncoherent
+    SnapshotDirectoryInitialized -> do
+      snapshotInfo <- getDirectorySnapshotInfo $ snapshotRoot s
+      case snapshotInfo of
         Left _ ->
           return DirectoryOutDated
         Right info -> do
-          update <- latestUpdate (repositoryUrl repo) (repositoryBranch repo)
+          update <- latestUpdate $ updatesUrl s
           return $
-            if update == Right (zonedTimeToUTC $ lastModificationCommitDate info)
+            if update == Right (lastModificationCommitDate info)
               then DirectoryUpToDate
               else DirectoryOutDated
```

Comment:
I think there is a problem here:

- `update` will be the date of the last commit to the snapshot branch.
- `lastModificationCommitDate` is from the snapshot metadata, and (if I'm not mistaken) is the date of the most recent advisory modification

Therefore, `update` is always newer than `lastModificationCommitDate`, and this expression always yields `DirectoryOutDated`.

I also think that querying the atom feed for the branch from GitHub is a very awkward way to determine whether there is a newer snapshot to retrieve.

GitHub tarballs are served with an ETag.  Could we store the ETag (on the client side).  When we want to check if a newer snapshot is available, perform a `HEAD` request with the `If-None-Match` header.  If the server responds `304 Not Modified`, the snapshot is fresh.  If `200`, the snapshot is out of date.  I tested this and it works as expected on GitHub.  This approach is also likely to be widely compatible with other HTTP servers, if we end up hosting the snapshot elsewhere.

I also tested `If-Modified-Since` and that **does not work** on GitHub.

Reply:
It seems reasonable, I'll rewrite it.